src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,411 @@
+                    yield (X509Certificate) cf.generateCertificate(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.X509_CRL -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509CRL) cf.generateCRL(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.RSA_PRIVATE_KEY -> {
+                    KeyFactory kf = getKeyFactory("RSA");
+                    yield kf.generatePrivate(
+                        RSAPrivateCrtKeyImpl.getKeySpec(decoder.decode(
+                            pem.pem())));
+                }
+                default -> pem;
+            };
+        } catch (GeneralSecurityException | IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given string.
+     *
+     * @param str a String containing PEM data.
+     * @return a {@code DEREncodable} generated from the PEM data.
+     * @throws IllegalArgumentException on error in decoding or if the PEM is
+     * unsupported.
+     * @throws NullPointerException when {@code str} is null.
+     */
+    public DEREncodable decode(String str) {
+        Objects.requireNonNull(str);
+        try {
+            return decode(new ByteArrayInputStream(
+                str.getBytes(StandardCharsets.ISO_8859_1)));
+        } catch (IOException e) {
+            // With all data contained in the String, there are no IO ops.
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given
+     * {@code InputStream}.
+     *
+     * <p>This method will read the {@code InputStream} until PEM data is
+     * found or until the end of the stream.  Non-PEM data in the
+     * {@code InputStream} before the PEM header will be ignored by the decoder.
+     * If only non-PEM data is found a {@link PEMRecord} is returned with that
+     * data.
+     *
+     * @param is InputStream containing PEM data.
+     * @return an {@code DEREncodable} generated from the data read.
+     * @throws IOException on IO error with the InputStream.
+     * @throws IllegalArgumentException on error in decoding or if the PEM is
+     * unsupported.
+     * @throws NullPointerException when {@code is} is null.
+     */
+    public DEREncodable decode(InputStream is) throws IOException {
+        Objects.requireNonNull(is);
+        PEMRecord pem = Pem.readPEM(is);
+        return decode(pem);
+    }
+
+    /**
+     * Decodes and returns the specified class for the given PEM string.
+     * {@code tClass} must extend {@link DEREncodable} and be an appropriate
+     * class for the PEM type.
+     *
+     * <p>
+     * {@code tClass} can be used to change the return type instance:
+     * <ul>
```

Comment:
This list feels more like programming advice. I would suggest either removing this, or putting it in the class description but not as a bulleted list, instead use complete sentences.